Apps/Admin/Server/Services/AgentAccessService.cs

```diff
@@ -0,0 +1,202 @@
+            _ = configuration;
+            (this.tokenUri, this.tokenRequest) = this.authDelegate.GetClientCredentialsAuth(AuthConfigSectionName);
+        }
+
+        /// <inheritdoc/>
+        [SuppressMessage("Globalization", "CA1308:Normalize strings to uppercase", Justification = "Not making security decisions based on the result of the normalization")]
+        public async Task<AdminAgent> ProvisionAgentAccessAsync(AdminAgent agent)
+        {
+            JwtModel jwtModel = this.authDelegate.AuthenticateAsSystem(this.tokenUri, this.tokenRequest);
+
+            agent.Username = agent.Username.ToLowerInvariant();
+            string userName = agent.Username + "@" + EnumUtility.ToEnumString<KeycloakIdentityProvider>(agent.IdentityProvider, true);
+            IEnumerable<RoleRepresentation> roles = await this.GetRoleRepresentationsAsync(agent.Roles).ConfigureAwait(true);
+
+            UserRepresentation user = new()
+            {
+                Username = userName,
+                Enabled = true,
+            };
+
+            try
+            {
+                await this.keycloakAdminApi.AddUserAsync(user, jwtModel.AccessToken).ConfigureAwait(true);
+            }
+            catch (Exception e) when (e is ApiException)
+            {
+                if (e.Message.Contains("409", StringComparison.InvariantCulture))
+                {
+                    this.logger.LogWarning(ErrorMessages.KeycloakUserAlreadyExists);
+                    throw new ProblemDetailsException(ExceptionUtility.CreateProblemDetails(ErrorMessages.KeycloakUserAlreadyExists, HttpStatusCode.Conflict, nameof(AgentAccessService)));
+                }
+
+                this.logger.LogError(e, "Keycloak API call failed");
+                throw;
+            }
+
+            List<UserRepresentation> createdUser = await this.keycloakAdminApi.GetUserAsync(userName, jwtModel.AccessToken).ConfigureAwait(true);
+            await this.keycloakAdminApi.AddUserRolesAsync(createdUser.First().UserId.ToString(), roles, jwtModel.AccessToken).ConfigureAwait(true);
+
+            return agent;
+        }
+
+        /// <inheritdoc/>
+        public async Task<IEnumerable<AdminAgent>> GetAgentsAsync(string queryString, int? resultLimit = 25)
+        {
+            const int firstRecord = 0;
+
+            JwtModel jwtModel = this.authDelegate.AuthenticateAsSystem(this.tokenUri, this.tokenRequest);
+
+            List<UserRepresentation> users = await this.keycloakAdminApi.GetUsersQueryAsync(queryString, firstRecord, resultLimit.GetValueOrDefault(), jwtModel.AccessToken).ConfigureAwait(true);
+
+            List<AdminAgent> adminAgents = new();
+            foreach (UserRepresentation user in users)
+            {
+                List<RoleRepresentation> userRoles = await this.keycloakAdminApi.GetUserRolesAsync(user.UserId.ToString(), jwtModel.AccessToken).ConfigureAwait(true);
+                string userName = user.Username.Split('@').First();
```

Comment:
Duplicate split? Also a potential null/not found exception.

Reply:
Fixing